vertical-pod-autoscaler/docs/proposals/crash_loop_handling/README.md

```diff
@@ -0,0 +1,196 @@
+  - 0.9 percentile for target CPU and Memory
+  - 0.5 for lower bound requests of CPU and Memory
+  - 0.95 for upper bound requests of CPU and Memory
+- For Target Estimator we only recommend Percentile Estimation with a safety margin of 15% by default
+- An exponential variation in the lower bound requests, upper bound requests which is determined by history of metrics monitored and pod life span. There is a "confidence" value on which the entire resource scaling is determined.
+- For lower bound and upper bound estimation a configurable safety margin (by default 15%) is applied
+- The current lower bound and upper bound estimation formula is as follows
+```
+scaledResource = originalResource * (1 + 1/confidence)^exponent
+
+confidence = (LastSampleTimeStart - FirstSampleTimeStart) / (no. of hours per day i.e 24)
+```
+
+Where,
+- **scaledResource**  : newly scaled integer value of the amount of the 
+                    resource  (CPU or Memory)
+- **originalResource**: current integer value of the amount of the
+                    resource (CPU or Memory)
+- **confidence**      : An non-negative integer number that heuristically measures 
+                    how much confidence the history aggregated in the AggregateContainerState provides. For a workload producing a steady stream of samples over N days at the rate of 1 sample per minute, this metric is equal to N. It is calcualted as the total count of samples and the time between the first and the last sample 
+- **exponent**        : A non-zero float exponent value which is determined by the 
+                    targetEstimator, lowedBoundEstimator or upperBoundEstimator. Each of these estimators are set through percentile estimations. targetEstimator is set to percentileEstimator while lowerBoundEstimator and upperBoundEstimator are set as confidence multipliers. For lowerBoundEstimator the exponent is set to -2 and for upperBoundEstimator it is set to +1
+
+## Proposal
+
+To address the concerns mentioned we are proposing this change in the recommendation engine of VPA recommender for autoscaler component.
+
+#### Scale Up:
+Doubling of resources based on current resource usage
+  1. Today we depend on the resource requests history from prometheus client for a history length of 8 days (default)
+  2. Instead, we could obtain current and past 10 minutes usage from prometheus client to get a nearest estimate of how much requests for resources is a pod making
+  3. In the event of a spike in resource, we can recommend the target estimator, requests as double the current resource usage. The lower bound and upper bound become the function of target estimator instead of directly recommending based on resource usage using percentile approach. Lower bound can be set to 0.5 times the target estimator value and upper bound can be set to 2 times the target estimator value.
```

Comment:
> The lower bound and upper bound become the function of target estimator instead of directly recommending based on resource usage using percentile approach

What's the motivation behind that?

If the idea is to speed up scale-up or scale-down, have you considered using different percentiles for lower and upper bound to faster trigger the vpa-updater?

Reply:
We want to make Lower Bound (LB) and Upper Bound (UB) as a function of target estimator to keep things simple really :) The entire percentile definition is confusing when it comes to recommending based on history. Since we are not using history, we want to make LB and UB a function of "something". That would be target estimator as we are determining that based on requests and usage of pod.

Reply:
So, are you saying with your proposal, you basically want to entirely drop the upstream recommendation mechanism based on historic data and only recommend based on the last 10min or similar?
This really doesn't sound like a good idea to me, TBH.

Reply:
@timebertt that is right. We are right now experimenting and also in the proposal itself showed a manual test run we made to depict how doubling really helps. Could you please let me know why this new proposal is not fully convincing? Also kindly suggest any alternate recommendations if so. Would really be glad to discuss further on this.

Reply:
In the past we have experienced quite some troubles with VPA, that give me a not so good gut feeling about just looking at the last 10min of resource usage:
- components scaling up too quickly when being used make a pretty bad UX: e.g. prometheus is often scaled up quite fasst when you throw a lot of queries at it, making it unavailable shortly after starting some investigation. This would happen even more often with your proposed approach, I assume
- components are having way too high requests and thereby we waste a lot of compute resources. Same here, I think when focus on spikes and double the requests very quickly and then prolong the delay until scale down, we will have even lower resource utilization than we already have
- scale-up/down events can be quite disruptive, especially for stateful components where we have to move volumes (which is slow and troublesome on some infrastructures). Me feeling is that we will have a lot more scaling happening if we only look at the last 10 mins of resource usage, which will make this situation even worse.

Having said that, all of these are just my expectations based on experiences over the last year or two.
I just wanted to raise them here, as you were looking for such feedback, as far as I understood.
Nevertheless, I'm open for you to prove me wrong on those assumptions and convince me that your proposals improves the mentioned problems.
I think it will definitely help here to have a personal discussion on this, so we can clarify questions and concerns more easily.